### PR TITLE
024 registration system

### PR DESCRIPTION
**Pull request by adamlelliott**

### CODE REVIEW

**Comment by adamlelliott** [comment]
API Changes: https://github.com/michaelkerrigan/ApiTeamB/pull/6

File: app/userdata.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+const mysql = require('mysql');
+const axios = require('axios');
+const dbconfig = require('../dbconfig.json');
+const util = require('util')
+
+const db = wrapDB(dbconfig)
+
+
+
+function wrapDB(dbconfig) {
```
Comment (shaunganley) [inline app/userdata.js]: this isn't used, we don't talk to the db using the frontend

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,4 @@
+dbconfig.json
```
Comment (shaunganley) [inline .gitignore]: we shouldn't need this file in the UI

File: app/userdata.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      release() {
+        return util.promisify(pool.releaseConnection)
+          .call(pool)
+      }
+    }
+  }
+  // Check this method 
+  exports.createUser = async (newUser) => {
+    try {
+      const usersResponse = await axios.post('http://localhost:8080/api/registeruser', { email: newUser.email, password: newUser.password, role: newUser.role })
```
Comment (shaunganley) [inline app/userdata.js]: you're passing the password in plaintext over http here, you need to encrypt either your transmission (using https) or your password before sending

File: app/views/login.html [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE html> 
+<html lang="en"> 
+<head> 
+<title>Login</title> 
+</head> 
+<body> 
+<h1>Registration successfull</h1>
```
Comment (shaunganley) [inline app/views/login.html]: typo

File: app/views/registration.html [hunk 5]
```diff
@@ -0,0 +1,81 @@
+{% from "govuk/components/table/macro.njk" import govukTable %}
+{% extends "layout_unbranded.html" %}
+{% from "govuk/components/select/macro.njk" import govukSelect %}
+{% from "govuk/components/button/macro.njk" import govukButton %}
+
+{% block pageTitle %}
+  {{ Registration }}
```
Comment (shaunganley) [inline app/views/registration.html]: do you need the curly brackets here?

File: dbconfig.json [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (shaunganley) [inline dbconfig.json]: can be removed

File: security/cert.key [hunk 7]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (shaunganley) [inline security/cert.key]: I'm not sure storing a private key in a repo is entirely secure

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: app/userdata.js (hunk 3 above)
Comment (adamlelliott) [reply app/userdata.js]: We've got the functionality for HTTPS, but it is using a locally generated key that hasn't been verified, would it be better to swap entirely to an encryption before transmission approach?

**Comment by adamlelliott** [comment]
Used BFG to purge any traces of the used keys from previous commits 😁

**Review by shaunganley — APPROVED** [review]

**Review by maxlilleyman — APPROVED** [review]

File: app/routes.js [hunk 8]
```diff
@@ -1,14 +1,36 @@
...
+router.post('/registeruser' , async (req, res) => {
+    let error = userValidator.validateUser(req.body)
+
+    if (error) {
+        res.locals.errormessage = error
+        return res.render('registration', req.body);
+    } else {
+        try {
+            var newUser = req.body;
+            await userdata.createUser(newUser);
```
Comment (maxlilleyman) [inline app/routes.js]: Could lines 22 and 23 be one line (CreateUser(req.body))

File: app/views/registration.html [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+          },
+          {
+            value: "admin",
+            text: "Admin"
+          }
+
+        ]
+
+
+      }) }}
```
Comment (maxlilleyman) [inline app/views/registration.html:68]: There is a lot of empty space here

File: app/routes.js (hunk 8 above)
Comment (adamlelliott) [reply app/routes.js]: Didn't spot that, cheers 😁

**Review by michaelkerrigan — APPROVED** [review]

File: app/validators/userValidator [hunk 10]
```diff
@@ -0,0 +1,21 @@
+const passwordRegex = RegExp(/^(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*]).{9,}$/)
+
+module.exports.validateUser = function (user) {
+    if (user.email == null || user.email.trim().length === 0) {
+        return "E-Mail must be provided"
+    }
+
+    if (!passwordRegex.test(user.password)) {
+        return "Password must be more than 8 characters, contain 1 upper case letter, 1 lower case letter, and a special character"
```
Comment (MatSwiKainos) [inline app/validators/userValidator:9]: Just like on the API side - error should also inform about need of a number inside a password:
return "Password must be more than 8 characters, contain 1 upper case letter, 1 lower case letter, 1 number and a special character"

**Pull request merged**